File: Algorithm/2023.12/wolfy916/알고리즘 테마정리/이진 탐색/보석 상자.py
'''
보석 상자 - 백준 2792, 실버1
분류 : 이진 탐색
'''
import sys, math

# [A] 입력함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [B] 이진 탐색 중간값 유효성 검사 함수
def check(N, mid, jewels):
    cnt = 0
    for jewel in jewels:
        cnt += math.ceil(jewel / mid)
        if cnt > N: return False
    return True

# [C] 이진탐색 함수
def binary_search(N, M, jewels):
    answer = 0
    left, right = 1, 10**9
    while left <= right:
        mid = (left + right) // 2
        if check(N, M, mid, jewels):
            right = mid - 1
            answer = mid
        else:
            left = mid + 1
    return answer

# [main]
if __name__ == '__main__':
    N, M = map(int, input().split())
    jewels = [int(input()) for _ in range(M)]

    print(binary_search(N, M, jewels))

File: Algorithm/2023.12/graywhales/보석상자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

/*
이분 탐색으로 jealous 가능한 값 찾기
 */
public class 보석상자 {
    static int N, M;
    static int[] arr;
    static int max = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String input = br.readLine();
        N = Integer.parseInt(input.split(" ")[0]);
        M = Integer.parseInt(input.split(" ")[1]);
        arr = new int[M];
        for (int i = 0; i < M; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            max = Math.max(max, arr[i]);
        }
        System.out.println(bs());
    }

    private static int bs() {
        int left = 1, right = max;
        while (left < right) {
            int mid = (left + right) / 2;
            if (check(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return right;
    }

    private static boolean check(int jealous) {
        int cnt = 0;
        for (int jewel : arr) {
            cnt += ((jewel + jealous - 1) / jealous);
        }
        return cnt <= N;
    }
}
